### PR TITLE
getGroupUsage function added

### LINKED ISSUE #439: Make NonCustodialSites selection better
At the moment NonCustodialSites is chosen randomly from available/allowed T1 sites.
We should change the default setting to make it empty unless specified in the campaign configuration (such as for premix library production to be sent to FNAL and CERN).

### PR DESCRIPTION
**Pull request by sharad1126**
Fixes #439 

#### Status
Completed | Tested

#### Description
Using the groupusage API mentioned by Nick to get the dest_bytes which will be subtracted from quota.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
N/A

#### External dependencies / deployment changes
it depends on https://cmsweb.cern.ch/phedex/datasvc/json/prod/groupusage?node=T1_UK_RAL_Disk&group=DataOps API where node changes on function call.

#### Mention people to look at PRs
@thongonary

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -2113,6 +2113,17 @@ def getNodeUsage(url, node):
...
+def getGroupUsage(url, node):
+    conn = make_x509_conn(url)
+    r1=conn.request("GET", '/phedex/datasvc/json/prod/groupusage?node=%s&group=DataOps'%node)
+    r2=conn.getresponse()
+    result = json.loads(r2.read())
+    for a in result['phedex']['node']:
+        for b in a.keys():
+            if b == 'group':
+                for e in a[b]:
+                    return (e["dest_bytes"])
```
[the remark this answers is not in the document]
Comment (sharad1126) [reply utils.py]: I tried that but that doesn't works.

**Comment by sharad1126** [comment]
> I think you need to modify the formula as Nick suggested as well.

yes, I will do that and commit.

the new formula now becomes available = int(float(quota))-int(getGroupUsage(url,site)) right?

File: utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (thongonary) [reply utils.py]: What does it mean it doesn't work? What error does it return?

[the remark this answers is not in the document]
Comment (sharad1126) [reply utils.py]:
```
>>> data['phedex']['node']['group']['dest_bytes']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: list indices must be integers, not str
```

[the remark this answers is not in the document]
Comment (thongonary) [reply utils.py]: OK so it's using a combination of list and dict. Fine.

File: utils.py [hunk 2]
```diff
@@ -2614,7 +2625,7 @@ def fetch_detox_info(self, talk=True, buffer_level=0.8, sites_space_override=Non
             #print site,self.queue.get(site,0)
             ## consider quota to be 80% of what's available
             queued_used = 0
-            available = int(float(quota)*buffer_level) - int(locked) - int(queued_used)
+            available = float(quota) - float(getGroupUsage(phedex_url,site))
```
Comment (thongonary) [inline utils.py]: Did you check if the value are in the same unit (GB/TB)?

Comment (sharad1126) [reply utils.py]: dest_bytes is in bytes but quota is in TB. I will change both to the same unit.

**Comment by sharad1126** [comment]
@thongonary apparently, it prints out something like:
```
quota: 2676
groupusage: 2318
quota: 833
groupusage: 2306
quota: 1174
groupusage: 1112
quota: 1819
groupusage: 1468
quota: 1650
groupusage: 1650
quota: 2000
groupusage: 3220
quota: 6248
groupusage: 5888
quota: 24
groupusage: 4
quota: 369
groupusage: 322
quota: 93
groupusage: 95
quota: 100
groupusage: 82
quota: 15
groupusage: None
```
I am not sure where this None comes from.

**Comment by thongonary** [comment]
Likely coming from your nested loops in `getGroupUsage` that doesn't meet the `if` requirement. Just print out each step in that function to debug.

**Comment by sharad1126** [comment]
I found the reason of the crash while running this. 
It goes through a site which has nothing inside the node and therefore the function returns a value that is None.
https://cmsweb.cern.ch/phedex/datasvc/json/prod/groupusage?node=T2_GR_Ioannina&group=DataOps

**Comment by sharad1126** [comment]
@thongonary I put a check for the length so cases like these sites return 0 and we can update quota-0. updated and tested. ready to merge.

**Comment by thongonary** [comment]
@sharad1126 please add the output size approximation in the equation as well.

**Comment by nsmith-** [comment]
According to PhEDEx, T2_GR_Ioannina has no data under the DataOps group presently: https://cmsweb.cern.ch/phedex/prod/Reports::SiteUsage?node=T2_GR_Ioannina#

**Comment by sharad1126** [comment]
> @sharad1126 please add the output size approximation in the equation as well.

@thongonary Isn't https://github.com/CMSCompOps/WmAgentScripts/blob/e7775aeeabc95036c1970fed7341c59cfc68eea1/utils.py#L7599 checking that already?

**Comment by thongonary** [comment]
Nope, that was regarding splitting for each job and has nothing to do with the output size of the whole output dataset.

**Comment by sharad1126** [comment]
so, how are we suppose to calculate the approximate output size of the dataset?

**Comment by thongonary** [comment]
Write a function to calculate it:
- If there is no input data, then take `SizePerEvents` x `RequestNumEvents` for each task and sum them over.
- If there is input data, then take `SizePerEvents` x number of events in the input dataset for each task and sum them over.

**Pull request closed without merging**